File: Mother Vertex[06 September 2023]/Solution.cpp
//  >> CPP CODE <<

class Solution {
public:
    // Function to find a Mother Vertex in the Graph.
    void dfs(int i, vector<int> adj[], vector<bool>& visited) {
        visited[i] = true;
        
        for (int v : adj[i]) {
            if (!visited[v]) {
                dfs(v, adj, visited);
            }
        }
    }

    int findMotherVertex(int V, vector<int>adj[]) {
        int lastNode = 0;
        vector<bool> visited(V, false);

        for (int i = 0; i < V; i++) {
            if (!visited[i]) {
                lastNode = i;
                dfs(i, adj, visited);
            }
        }
        
        fill(visited.begin(), visited.end(), false);
        
        dfs(lastNode, adj, visited);
        
        for (int i = 0; i < V; i++) {
            if (!visited[i]) {
                return -1;
            }
        }
        
        return lastNode;
    }
};

File: Mother Vertex[06 September 2023]/Solution.java
//  >> JAVA CODE <<

class Solution
{
    //Function to find a Mother Vertex in the Graph.
    public void dfs(int i,ArrayList<ArrayList<Integer>> adj,boolean[] visited)
    {
        visited[i]=true;
        
        for(var v:adj.get(i))
        {
            if(visited[v]==false)
                dfs(v,adj,visited);
        }
    }
    public int findMotherVertex(int V, ArrayList<ArrayList<Integer>>adj)
    {
        int lastNode=0;
        boolean[] visited=new boolean[V];
        for(int i=0;i<V;i++)
        {
            if(visited[i]==false)
            {
                lastNode=i;
                dfs(i,adj,visited);
            }
        }
        
        Arrays.fill(visited,false);
        
        dfs(lastNode,adj,visited);
        
        for(int i=0;i<V;i++)
        {
            if(visited[i]==false) return -1;
        }
        
            return lastNode;
    }
}